language_service/src/state.rs

```diff
@@ -0,0 +1,465 @@
+    state: Rc<RefCell<CompilationState>>,
+    /// Workspace-wide configuration settings. These can include compiler settings that
+    /// affect error checking and other language server behavior.
+    ///
+    /// Some settings can be set both at the compilation scope and at the workspace scope.
+    /// Compilation-scoped settings take precedence over workspace-scoped settings.
+    configuration: Configuration,
+    /// Documents that we have previously published errors about. We need to
+    /// keep track of this so we can clear errors from them when documents are removed
+    /// from a compilation or when a recompilation occurs.
+    documents_with_errors: FxHashSet<DocumentUri>,
+    /// Callback which will receive diagnostics (compilation errors)
+    /// whenever a (re-)compilation occurs.
+    diagnostics_receiver: Box<dyn Fn(DiagnosticUpdate) + 'a>,
+    /// Callback which lets the service read a file from the target filesystem
+    pub(crate) read_file_callback: AsyncFunction<'a, String, (Arc<str>, Arc<str>)>,
+    /// Callback which lets the service list directory contents
+    /// on the target file system
+    pub(crate) list_directory: AsyncFunction<'a, String, Vec<JSFileEntry>>,
+    /// Fetch the manifest file for a specific path
+    get_manifest: AsyncFunction<'a, String, Option<qsc_project::ManifestDescriptor>>,
+}
+
+impl<'a> CompilationStateUpdater<'a> {
+    pub fn new(
+        state: Rc<RefCell<CompilationState>>,
+        diagnostics_receiver: impl Fn(DiagnosticUpdate) + 'a,
+        read_file: impl Fn(String) -> Pin<Box<dyn Future<Output = (Arc<str>, Arc<str>)>>> + 'a,
+        list_directory: impl Fn(String) -> Pin<Box<dyn Future<Output = Vec<JSFileEntry>>>> + 'a,
+        get_manifest: impl Fn(String) -> Pin<Box<dyn Future<Output = Option<qsc_project::ManifestDescriptor>>>>
+            + 'a,
+    ) -> Self {
+        Self {
+            state,
+            configuration: Configuration::default(),
+            documents_with_errors: FxHashSet::default(),
+            diagnostics_receiver: Box::new(diagnostics_receiver),
+            read_file_callback: Box::new(read_file),
+            list_directory: Box::new(list_directory),
+            get_manifest: Box::new(get_manifest),
+        }
+    }
+
+    /// Updates the workspace configuration. If any compiler settings are updated,
+    /// a recompilation may be triggered, which will result in a new set of diagnostics
+    /// being published.
+    pub fn update_configuration(&mut self, configuration: &WorkspaceConfigurationUpdate) {
+        trace!("update_configuration: {configuration:?}");
+
+        let need_recompile = self.apply_configuration(configuration);
+
+        // Some configuration options require a recompilation as they impact error checking
+        if need_recompile {
+            self.recompile_all();
+        }
+    }
+
+    pub(super) async fn update_document(&mut self, uri: &str, version: u32, text: &str) {
```

Comment:
This method is an exmple of the project special casing we discussed. Loading a project from a path should always return a project/manifest regardless of whether there is a project. Then all logic is merged.

Reply:
Forwarding this comment to @sezna  to take into account as he reworks the project load / open buffer merge code path. I didn't want to touch any of the logic here since he's working off my branch.